File: types/assumptions.ts
export interface AssumptionsSnapshot {
  ceiling_height_ft: number;
  waste_factor: number;
  sheet_width_ft: number;
  sheet_length_ft: number;
  drywall_unit_cost_usd: number | null;
  markup_pct: number | null;
  tax_rate_pct: number | null;
}

export interface AssumptionOverride {
  key: keyof AssumptionsSnapshot;
  value: number;
  locked: boolean;
  reason: string;
  set_at: string;
}

export const DEFAULT_ASSUMPTIONS: AssumptionsSnapshot = {
  ceiling_height_ft: 9.0,
  waste_factor: 0.15,
  sheet_width_ft: 4.0,
  sheet_length_ft: 12.0,
  drywall_unit_cost_usd: null,
  markup_pct: null,
  tax_rate_pct: null,
};

File: backend/src/api/routes/assumptions.py
"""Assumptions API — per-project, per-page assumption overrides."""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter
from pydantic import BaseModel

from src.api.assumptions_store import (
    load_assumptions,
    save_assumptions,
    load_locked_keys,
    save_locked_keys,
)
from src.schemas.estimate import AssumptionsSnapshot

router = APIRouter(prefix="/api/assumptions", tags=["assumptions"])


class AssumptionsPatch(BaseModel):
    ceiling_height_ft: Optional[float] = None
    waste_factor: Optional[float] = None
    sheet_width_ft: Optional[float] = None
    sheet_length_ft: Optional[float] = None
    drywall_unit_cost_usd: Optional[float] = None
    markup_pct: Optional[float] = None
    tax_rate_pct: Optional[float] = None


class LockPatch(BaseModel):
    locked_keys: list[str]


@router.get("/{project_id}", response_model=AssumptionsSnapshot)
def get_assumptions(project_id: str, page: int = 1) -> AssumptionsSnapshot:
    return load_assumptions(project_id, page)


@router.patch("/{project_id}", response_model=AssumptionsSnapshot)
def patch_assumptions(
    project_id: str, body: AssumptionsPatch, page: int = 1
) -> AssumptionsSnapshot:
    current = load_assumptions(project_id, page)
    updated = current.model_dump()
    for key, value in body.model_dump(exclude_none=True).items():
        updated[key] = value
    snapshot = AssumptionsSnapshot(**updated)
    save_assumptions(project_id, page, snapshot)
    return snapshot


@router.get("/{project_id}/locks", response_model=list[str])
def get_locks(project_id: str, page: int = 1) -> list[str]:
    return load_locked_keys(project_id, page)


@router.put("/{project_id}/locks", response_model=list[str])
def set_locks(project_id: str, body: LockPatch, page: int = 1) -> list[str]:
    save_locked_keys(project_id, page, body.locked_keys)
    return body.locked_keys
